### ISSUE TITLE
How to impove small object detection

### ISSUE THREAD

**Issue by ChungChinKei** [AlexeyAB/darknet#3198]
@AlexeyAB Hi,

Thank you so much for sharing this project.
I am trying to detect some small defects on Metal surface.

**Some information about dataset:**
Only 1 class.
Grayscale images.
200 images for training(100 images of positive data and 100 negative).
50 images for validation(all positive).
Images resolution is 1600 x 1200.
objects size is about 30 x 30.(even smaller)
yolov3-tiny_3l.cfg with generated anchors.

**Positive sample**
![NG(4)](https://user-images.githubusercontent.com/44358330/58064192-b9518280-7bb3-11e9-8460-6616463f5265.jpg)
![NG(7)](https://user-images.githubusercontent.com/44358330/58064359-4eed1200-7bb4-11e9-95ba-bd9cd6ca109a.jpg)

**Negative sample**
![OK12](https://user-images.githubusercontent.com/44358330/58064132-860ef380-7bb3-11e9-8769-202a7a675d81.jpg)

Could you give me some advise to impove mAP？
![chart_1024X768_8000](https://user-images.githubusercontent.com/44358330/58063652-b05fb180-7bb1-11e9-928b-02e5b503336b.png)

**Some details of cfg:**
[net]
batch=64
subdivisions=32
width=1024
height=768
channels=3
momentum=0.9
decay=0.0005
angle=0
saturation = 1
exposure = 1.5
hue= 0

learning_rate=0.005
burn_in=1000
max_batches = 10000
policy=steps
steps=8000,9000
scales=.1,.1

[yolo]
mask = 6,7,8
anchors = 21, 19,  29, 19,  27, 23,  30, 29,  36, 24,  48, 27,  40, 35,  77, 22,  56, 42
classes=1
num=9
jitter=.3
ignore_thresh = .7
truth_thresh = 1
random=1

**Comment by jonasmuriboe** [comment]
I would recommend using the full size YOLOv3 architecture like in yolov3.cfg, or maybe even yolov3-spp.cfg as this has an even higher mAP.  You should try the two following modifications as well as I have personally had a drastic improvement in performance by applying them to a network that attempts to find very small objects, typically in the same size-range as you.

> for training for small objects (smaller than 16x16 after the image is resized to 416x416) - set layers = -1, 11 instead of https://github.com/AlexeyAB/darknet/blob/6390a5a2ab61a0bdf6f1a9a6b4a739c16b36e0d7/cfg/yolov3.cfg#L720 and set stride=4 instead of https://github.com/AlexeyAB/darknet/blob/6390a5a2ab61a0bdf6f1a9a6b4a739c16b36e0d7/cfg/yolov3.cfg#L717

As well as this

> recalculate anchors for your dataset for width and height from cfg-file: darknet.exe detector calc_anchors data/obj.data -num_of_clusters 9 -width 416 -height 416 then set the same 9 anchors in each of 3 [yolo]-layers in your cfg-file. But you should change indexes of anchors masks= for each [yolo]-layer, so that 1st-[yolo]-layer has anchors larger than 60x60, 2nd larger than 30x30, 3rd remaining. Also you should change the filters=(classes + 5)*<number of mask> before each [yolo]-layer. If many of the calculated anchors do not fit under the appropriate layers - then just try using all the default anchors.

**Comment by AlexeyAB** [comment]
@ChungChinKei Hi,

1. Try to train with default anchors.

2. Also you can try to train with SGDR:
change these lines: https://github.com/AlexeyAB/darknet/blob/9c9344a1ff4541499c8f69ea997be5145b4e1de3/cfg/yolov3-tiny_3l.cfg#L21-L23
to these lines: 
```
policy=sgdr
sgdr_cycle=1000
sgdr_mult=2
steps=4000,6000,8000,9000
#scales=1, 1, 0.1, 0.1
```

----

Also you can try to use one of new models (without `LSTM` in filename - since LSTM is only for detection on video): https://github.com/AlexeyAB/darknet/issues/3114#issuecomment-494148968
* tiny: https://github.com/AlexeyAB/darknet/files/3199622/yolo_v3_tiny_pan.cfg.txt
* or full: https://github.com/AlexeyAB/darknet/files/3199657/yolo_v3_spp_pan.cfg.txt

But they are supported **only by this repository.**

**Comment by ChungChinKei** [comment]
@AlexeyAB @jonasmuriboe 

Thanks for your reply and suggestions .
But some errors occurred when I used the new PANet models.
How can I fix it?
![errors](https://user-images.githubusercontent.com/44358330/58219643-3c9edf80-7d3e-11e9-9b37-03c86a58c130.png)

**Comment by SHADOWMOOON** [comment]
@AlexeyAB  Thanks for the great work!
I wonder Why is the calculated anchors less fit than the default anchors. I've try to set the anchors at each mask layer(0.1.2/3.4.5/.6.7.8),but the come out was very unsatisfied.

**Comment by AlexeyAB** [comment]
@SHADOWMOOON 

> How can I fix it?

You must use the latest versions of this repository: https://github.com/AlexeyAB/darknet

> If I do this wrong,how can I use the calculated anchors correctly?Could you please give me some advisement.

You must change the masks (or you must change only some anchors): https://github.com/AlexeyAB/darknet#how-to-improve-object-detection

> recalculate anchors for your dataset for width and height from cfg-file: darknet.exe detector calc_anchors data/obj.data -num_of_clusters 9 -width 416 -height 416 then set the same 9 anchors in each of 3 [yolo]-layers in your cfg-file. **But you should change indexes of anchors masks= for each [yolo]-layer, so that 1st-[yolo]-layer has anchors larger than 60x60, 2nd larger than 30x30, 3rd remaining.** Also you should change the filters=(classes + 5)*<number of mask> before each [yolo]-layer. If many of the calculated anchors do not fit under the appropriate layers - then just try using all the default anchors.

**Comment by SHADOWMOOON** [comment]
@AlexeyAB 
what do you mean to change the masks,Could you give  an example?

**Comment by duohappy** [comment]
@SHADOWMOOON 
refer to 
[https://github.com/AlexeyAB/darknet/issues/3372](https://github.com/AlexeyAB/darknet/issues/3372)
[https://github.com/AlexeyAB/darknet/issues/2960](https://github.com/AlexeyAB/darknet/issues/2960)

**Comment by sadimoodi** [comment]
> @ChungChinKei

did Alexy suggesiton work for you? i am doing something similiar to you so it would help me if you have any further suggestions, 
thanks
